crates/typst-ide/src/definition.rs

```diff
@@ -0,0 +1,243 @@
+    });
+
+    if paths.is_empty() {
+        return src;
+    }
+
+    None
+}
+
+/// A definition of some item.
+#[derive(Debug, Clone)]
+pub struct Definition {
+    /// The name of the definition.
+    pub name: EcoString,
+    /// The kind of the definition.
+    pub kind: DefinitionKind,
+    /// A possible instance of the definition.
+    pub value: Option<Value>,
+    /// The source span of the entire definition.
+    pub span: Span,
+    /// The span of the name of the definition.
+    pub name_span: Span,
+}
+
+impl Definition {
+    fn with_kind(self, kind: DefinitionKind) -> Self {
+        Self { kind, ..self }
+    }
+
+    fn item(name: &EcoString, span: Span, name_span: Span, value: Option<Value>) -> Self {
+        let kind = value
+            .as_ref()
+            .and_then(|e| {
+                matches!(e, Value::Func(..)).then_some(DefinitionKind::Function)
+            })
+            .unwrap_or(DefinitionKind::Variable);
+
+        Self { name: name.clone(), kind, value, span, name_span }
+    }
+
+    fn module(module: &Module) -> Self {
+        Definition {
+            name: EcoString::new(),
+            kind: match module.file_id() {
+                Some(file_id) => DefinitionKind::ModulePath(file_id),
+                None => DefinitionKind::Module(module.clone()),
+            },
+            value: Some(Value::Module(module.clone())),
+            span: Span::detached(),
+            name_span: Span::detached(),
+        }
+    }
+}
+
+/// A kind of item that is definition.
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub enum DefinitionKind {
+    /// ```plain
+    /// let foo;
+    ///     ^^^
+    /// ```
+    Variable,
+    /// ```plain
+    /// let foo(it) = it;
+    ///     ^^^
+    /// ```
+    Function,
+    /// ```plain
+    /// import ("fo" + "o.typ")
+    ///          ^^^^^^^^
+    /// ```
+    ///
+    /// IDE will always resolve a path instead of a module whenever possible.
+    /// This allows resolving a module containing errors.
+    ModulePath(FileId),
+    /// ```plain
+    /// import calc: *
+    ///        ^^^^
+    /// ```
+    ///
+    /// Some modules are not associated with a file, like the built-in modules.
+    Module(Module),
+    /// ```plain
+    /// <foo>
+    ///  ^^^
+    /// ```
+    Label,
+}
+
+fn find_let_binding(source: &Source, name_span: Span) -> Span {
+    let node = LinkedNode::new(source.root());
+    std::iter::successors(node.find(name_span).as_ref(), |n| n.parent())
+        .find(|n| matches!(n.kind(), SyntaxKind::LetBinding))
+        .map(|s| s.span())
+        .unwrap_or_else(Span::detached)
+}
+
+#[cfg(test)]
+mod tests {
+    use typst::foundations::Value;
+    use typst::syntax::{Side, Span};
+
+    use super::{definition, Definition};
+    use crate::tests::TestWorld;
+
+    fn var(text: &str, value: bool) -> Option<Definition> {
+        Some(Definition::item(
+            &text.into(),
+            Span::detached(),
+            Span::detached(),
```

Comment:
why are the test spans all detached?

Reply:
They are not used, and I have no idea on test spans.